src/segment_reshape/geometry/replace_segment.py

```diff
@@ -17,76 +17,212 @@
 #  along with segment-reshape-qgis-plugin. If not, see <https://www.gnu.org/licenses/>.
 
+from contextlib import contextmanager
+from contextvars import ContextVar
 from dataclasses import dataclass
-from typing import Any, List
+from typing import Iterator, List, Set, Union
 
-from qgis.core import QgsFeature, QgsGeometry, QgsLineString, QgsPoint, QgsVectorLayer
+from qgis.core import (
+    QgsFeature,
+    QgsGeometry,
+    QgsLineString,
+    QgsPoint,
+    QgsVectorLayer,
+    QgsWkbTypes,
+)
+
+from segment_reshape.utils import clone_geometry_safely
+
+
+class GeometryTransformationError(Exception):
+    pass
 
 
 @dataclass
-class SegmentCommonFeature:
+class SegmentCommonPart:
     layer: QgsVectorLayer
     feature: QgsFeature
-    from_vertex_id: Any
-    to_vertex_id: Any
-    reversed: bool
+    vertex_indices: List[int]
+    is_reversed: bool
 
 
 @dataclass
-class SegmentEndPointFeature:
+class SegmentEdge:
     layer: QgsVectorLayer
     feature: QgsFeature
-    vertex_id: Any
+    vertex_index: int
     is_start: bool
 
 
-def make_edits_for_new_segment(
-    common_features: List[SegmentCommonFeature],
-    end_point_features: List[SegmentEndPointFeature],
-    new_segment: QgsLineString,
-) -> None:
-    pass
+_current_edit_command_layers: ContextVar[List[QgsVectorLayer]] = ContextVar(
+    "current_edit_command_layers"
+)
+_set_editable_layer_ids: ContextVar[Set[str]] = ContextVar("set_editable_layer_ids")
 
-    # begineditcommand
 
-    # _replace_segments
-    # _replace_vertices
+@contextmanager
+def _wrap_all_edit_commands() -> Iterator[None]:
+    layers: List[QgsVectorLayer] = []
+    set_editable_ids: Set[str] = set()
+    token = _current_edit_command_layers.set(layers)
+    editable_ids_token = _set_editable_layer_ids.set(set_editable_ids)
+    try:
+        yield
+        for layer in layers:
+            layer.endEditCommand()
+    except GeometryTransformationError as e:
+        for layer in layers:
+            layer.destroyEditCommand()
+            if layer.id() in set_editable_ids:
+                layer.rollBack()
+        raise e
+    finally:
+        _current_edit_command_layers.reset(token)
+        _set_editable_layer_ids.reset(editable_ids_token)
 
-    # endeditcommand
 
+def _set_editable_and_begin_edit_command_once(layer: QgsVectorLayer) -> None:
+    if layer.isEditCommandActive():
+        return
 
-def _replace_segments(
-    to_replace: List[SegmentCommonFeature], new_segment: QgsLineString
+    if not layer.isEditable():
+        if not layer.startEditing():
+            raise GeometryTransformationError(f"could not start editing on {layer}")
+        _set_editable_layer_ids.get().add(layer.id())
+
+    layer.beginEditCommand("Reshape segment")
+    _current_edit_command_layers.get().append(layer)
+
+
+def make_edits_for_new_segment(
+    segment_common_parts: List[SegmentCommonPart],
+    segment_edges: List[SegmentEdge],
+    reshaped_segment: Union[QgsPoint, QgsLineString],
```

Comment:
I think segment cannot be point (by definition segment has 2 end points) so this parameter name is a bit misleading? Could this be renamed as reshaped_segment_or_moved_point etc. or have 2 separate optional parameters (though I guess then it must be checked that excatly one of them has a value)? With same logic, the function name is also a bit misleading. Maybe change the function name to to just make_edits?

Reply:
Well in theory this code now support a case where a common segment is collapsed to a single point (ie. new segment is a single point), and reverse case where a single point is expanded into a common segment, so maybe a rename would be better.

Maybe "make_edits_for_reshape" is suitable, then we have "common_parts" and "edges", and a "reshape_geometry" for this generic api.